agent/rpc/peering/service.go

```diff
@@ -677,9 +683,14 @@ func (s *Service) HandleStream(req HandleStreamRequest) error {
 			if !s.Backend.IsLeader() {
 				// we are not the leader anymore so we will hang up on the dialer
 
-				// TODO(peering): in the future we want to indicate the address of the leader server as a message to the dialer (best effort, non blocking)
 				logger.Error("node is not a leader anymore; cannot continue streaming")
-				return grpcstatus.Error(codes.FailedPrecondition, "node is not a leader anymore; cannot continue streaming")
+				var merr *multierror.Error
+				if err := s.sendLeaderAddrMsg(req.Stream, status, logger); err != nil {
+					merr = multierror.Append(merr, err)
+				}
+
+				merr = multierror.Append(grpcstatus.Error(codes.FailedPrecondition, "node is not a leader anymore; cannot continue streaming"), merr)
+				return merr.ErrorOrNil()
```

Comment:
I think this implementation right now can be streamlined by either treating "is not leader" state as:
1. A regular replication message (similar to `Terminated`) and don't return errors but assume client will handle the LeaderAddress message gracefully; or
2. An error state where we send a `grpcstatus.Error` with `grpcstatus.Status` [holding Details about the error](https://cloud.google.com/apis/design/errors#error_model)

Doing both feels redundant and prone to more bugs since the client needs to handle 2 messages instead of one

Reply:
👍🏼 I'm with you there. Doing both kinda seems wrong.

I've been leaning towards `2` -- packing things into the error itself.  To me, sending an error to the dialer indicates that something is wrong and it needs to handle that.

do you (or others) have a preference?

Reply:
I don't think we've really set a precedence for `2` yet but it might be worth ironing things out. The downside is we'll have to add intricate error-handling logic on the receiving side. Any thoughts @freddygv @rboyer

Reply:
I like either way. Once we get server-list replication going, we might end up refactoring this anyway. At that point it would be good to replicate the list of server addresses with an annotation that says which one is currently leader.

If we had that, then what this "last send before error" would likely be better off doing is just re-issuing one of those normal replication messages instead, and then maybe issue a gRPC plain error here and let the receiving side use the replicated data to update its own choice of servers to contact.

Reply:
sounds like we are going with the `2`/ `error` approach -- I will update the PR with that 🤝

Reply:
We should have the leader address baked into the error details now.